garden_ai/utils.py

```diff
@@ -196,3 +201,105 @@ def f_of_g(*args, **kwargs):
     )
     f_of_g.__name__ = f.__name__ + "_COMPOSED_WITH_" + g.__name__
     return f_of_g
+
+
+def mint_doi(
+    title: str,
+    authors: List[str],
+    contributors: List[str],
+    year: str,
+    description: str = "",
+    publisher: str = "thegardens.ai",
+    resourceType: str = "AI/ML",
+    test: bool = True,
+):
+    """Request a "findable" DOI from DataCite
+
+    Note that this expects environment variables 'DATACITE_REPOSITORY_ID', 'DATACITE_PASSWORD' and
+    'DOI_PREFIX' to be set correctly in order to authenticate with DataCite.
+
+    Parameters
+    ----------
+    title : str
+        Title that should appear in citation
+    authors : list[str]
+    contributors: list[str]
+    year : str
+        YYYY
+    description: str
+        free-text description of the Digital Object being made Identifiable
+    publisher : str
+        "Garden" seems like a sane default?
+    resourceType : str
+        "AI/ML" by default -- this is allowed to be free text by DataCite but is advised to be kept short
+    test : bool
+        Whether to hit the DataCite 'test' API or the real one - currently only the former should be
+        used as it's the only one implemented.
+
+    Raises
+    ------
+    HTTPError
+        if one is encountered by `requests.Response.raise_for_status`
+
+    NotImplementedError
+        if given `test=False`
+    """
+    if not test:
+        datacite_url = "https://api.datacite.org/dois"
+        prefix = "10.26311"
+        raise NotImplementedError
+    else:
+        datacite_url = "https://api.test.datacite.org/dois"
+        prefix = "10.23677"
+
+    header = {"Content-Type": "application/vnd.api+json"}
+    payload = {
+        "data": {
+            "type": "dois",
+            "attributes": {
+                "event": "publish",
+                "prefix": prefix,
+                "creators": [{"name": author_name} for author_name in authors],
+                "contributors": [
+                    {"name": contrib_name, "contributorType": "Other"}
+                    for contrib_name in contributors
+                ],
+                "descriptions": [
+                    {"description": description, "descriptionType": "Other"}
+                ],
+                "titles": [{"title": title}],
+                "publisher": publisher,
+                "publicationYear": year,
+                "types": {
+                    "resourceTypeGeneral": "Software",
+                    "resourceType": resourceType,
+                },
+                "url": "http://thegardens.ai",
+                "schemaVersion": "http://datacite.org/schema/kernel-4",
+            },
+        }
+    }
+
+    try:
+        DATACITE_REPOSITORY_ID = os.environ["DATACITE_REPOSITORY_ID"]
+        DATACITE_PASSWORD = os.environ["DATACITE_PASSWORD"]
```

Comment:
This is something that's awkward for client vs backend, so consider this a note for the second pass of this that lives in the backend. (And is quite possibly something you're already thinking about)

Ideally when there's something wrong with the app config, you want to fail as quickly as possible. When this gets ported to the backend, this should trigger some alarms in our logs at app startup so that we can know there's an issue ASAP (even if a user doesn't hit this part of the code)